src/rt/array/construction.d

```diff
@@ -0,0 +1,165 @@
+
+    debug(PRINTF) printf("_d_arrayctor(to = %p,%d, from = %p,%d) size = %d\n", from.ptr, from.length, to.ptr, to.length, T.tsize);
+
+    auto element_size = T.sizeof;
+
+    void[] vFrom = (cast(void*)from.ptr)[0..from.length];
+    void[] vTo = (cast(void*)to.ptr)[0..to.length];
+    enforceRawArraysConformable("initialization", element_size, vFrom, vTo, false);
+
+    size_t i;
+    try
+    {
+        for (i = 0; i < to.length; i++)
+        {
+            auto elem = cast(Unqual!T*)&to[i];
+            // Copy construction is defined as bit copy followed by postblit.
+            memcpy(elem, &from[i], element_size);
+
+            static if (__traits(hasMember, T, "__xpostblit") &&
+                // Bugzilla 14746: Check that it's the exact member of S.
+                __traits(isSame, T, __traits(parent, elem.__xpostblit)))
+                elem.__xpostblit();
+        }
+    }
+    catch (Exception o)
+    {
+        /* Destroy, in reverse order, what we've constructed so far
+        */
+        while (i--)
+        {
+            auto elem = cast(Unqual!T*)&to[i];
+            static if (__traits(hasMember, T, "__xdtor") &&
+                // Bugzilla 14746: Check that it's the exact member of S.
+                __traits(isSame, T, __traits(parent, elem.__xdtor)))
+                try
+                    elem.__xdtor();
+                catch (Exception) {}
+        }
+
+        throw o;
+    }
+
+    return to;
+}
+
+@safe unittest
+{
+    int counter;
+    struct S
+    {
+        int val;
+        this(this)
+        {
+            counter++;
+        }
+    }
+
+    S[4] arr1;
+    S[4] arr2 = [S(0), S(1), S(2), S(3)];
+    _d_arrayctor(arr1, arr2);
+
+    assert(counter == 4);
+    assert(arr1 == arr2);
+}
+
+/**
+ * Do construction of an array.
+ *      ti[count] p = value;
+ * Params:
+ *  p = what array to initialize
+ *  value = what data to construct the array with
+ * Bugs:
+ *  The functions `@safe` safety is faked.
+ */
+void _d_arraysetctor(T)(scope T[] p, scope ref T value) @trusted
+{
+    import core.stdc.string : memcpy;
+    import core.internal.traits : Unqual;
+    size_t walker;
+    auto element_size = T.sizeof;
+
+    try
+    {
+        foreach (i; 0 .. p.length)
+        {
+            auto elem = cast(Unqual!T*)&p[walker];
+            // Copy construction is defined as bit copy followed by postblit.
+            memcpy(elem, &value, element_size);
+            static if (__traits(hasMember, T, "__xpostblit") &&
+                // Bugzilla 14746: Check that it's the exact member of S.
+                __traits(isSame, T, __traits(parent, elem.__xpostblit)))
+            elem.__xpostblit();
+            walker++;
+        }
+    }
+    catch (Exception o)
+    {
+        // Destroy, in reverse order, what we've constructed so far
+        while (walker > 0)
+        {
+            walker--;
+            auto elem = cast(Unqual!T*)&p[walker];
+            static if (__traits(hasMember, T, "__xdtor") &&
```

Comment:
Should this code be calling `object.destroy` instead?

Reply:
I changed it `object.destroy` and `object._postblitRecurse` for the postblit.